### PR TITLE
FINERACT-2117: Early and late payment - In advance payment strategy: …

### PR DESCRIPTION
**Pull request by alberto-art3ch**
…Adjust last, unpaid period

## Description

Progressive interest bearing loan product shall support: Early and late payment situations where interest recalculation is enabled.

In case of an early payment (in advance payment) we have different strategies that can be used. 

One of these strategy should be:
- Adjust last, unpaid period

The rule is simple, in case of an in advance payment the EMI of the last, unpaid period to be amended. 

Why last? 
We would like it to be reflected only at the end of the loan term, dont wanna change the EMI of the rest of the periods. 

[FINERACT-2117](https://issues.apache.org/jira/browse/FINERACT-2117)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractCumulativeLoanScheduleGenerator.java [hunk 1]
```diff
@@ -1552,6 +1552,8 @@ private Money applyEarlyPaymentStrategy(final LoanApplicationTerms loanApplicati
                 // until this
                 // amount becomes zero
                 break;
+                case ADJUST_LAST_UNPAID_PERIOD:
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractCumulativeLoanScheduleGenerator.java]: For cumulative loan this option shall not be available!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractCumulativeLoanScheduleGenerator.java [hunk 2]
```diff
@@ -2687,6 +2689,8 @@ private Money fetchEarlyPaidAmount(final Money principalPortion, final Money pri
...
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractCumulativeLoanScheduleGenerator.java]: For cumulative loan this option shall not be available!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java [hunk 3]
```diff
@@ -345,6 +346,22 @@ public void validateForCreate(final JsonCommand command) {
         // Validating whether the processor is existing
         loanRepaymentScheduleTransactionProcessorFactory.determineProcessor(transactionProcessingStrategyCode);
 
+        String loanScheduleProcessingType = LoanScheduleProcessingType.HORIZONTAL.name();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: Why did you move these checks?

Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: Done! restored

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractCumulativeLoanScheduleGenerator.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractCumulativeLoanScheduleGenerator.java]: Done! removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java [hunk 4]
```diff
@@ -1042,7 +1043,27 @@ private void validateInterestRecalculationParams(final JsonElement element, fina
...
+                baseDataValidator.reset().parameter(LoanProductConstants.rescheduleStrategyMethodParameterName).failWithCode(
+                        "reschedule.strategy.method.not.supported.for.loan.schedule.type.cumulative",
+                        "Adjust last, unpaid period is only supported for Progressive loan schedule type");
+            }
+
+            if (LoanScheduleType.PROGRESSIVE.equals(LoanScheduleType.valueOf(loanScheduleType))
+                    && !LoanRescheduleStrategyMethod.ADJUST_LAST_UNPAID_PERIOD.equals(loanRescheduleStrategyMethod)) {
+                baseDataValidator.reset().parameter(LoanProductConstants.rescheduleStrategyMethodParameterName).failWithCode(
+                        "reschedule.strategy.method.not.supported.for.loan.schedule.type.progressive",
+                        "Reschedule strategy type is not supported for Progressive loan schedule type");
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: Would you mind to rework the error message to contains which reschedule strategy is not supported?
Example:
- "Reduce EMI amount" reschedule strategy type is not supported for Progressive loan schedule type

I would be more descriptive if we are telling which strategy is not supported

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionAccrualActivityPostingTest.java [hunk 5]
```diff
@@ -1016,6 +1016,7 @@ private PostLoanProductsRequest loanProductsRequestPin30InterestDecliningBalance
                 .isEqualAmortization(false)//
                 .interestCalculationPeriodType(0)//
                 .transactionProcessingStrategyCode("mifos-standard-strategy")//
+                .loanScheduleType(LoanScheduleType.CUMULATIVE.toString())//
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionAccrualActivityPostingTest.java]: I dont think this is needed. This field is optional...

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 6]
```diff
@@ -5095,6 +5095,41 @@ public void uc147b() {
...
+    // period
+    @Test
+    public void uc148a() {
+        runAt("23 March 2024", () -> {
+            final Integer rescheduleStrategyMethod = 4; // Adjust last, unpaid period
+            PostLoanProductsRequest loanProduct = createOnePeriod30DaysPeriodicAccrualProductWithAdvancedPaymentAllocationAndInterestRecalculation(
+                    (double) 80.0, rescheduleStrategyMethod);
+
+            PostLoanProductsResponse loanProductResponse = loanProductHelper.createLoanProduct(loanProduct);
+            assertNotNull(loanProductResponse);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:5111]: Can you check whether the correct value was set?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 7]
```diff
@@ -5095,6 +5095,41 @@ public void uc147b() {
...
+            PostLoanProductsRequest loanProduct = createOnePeriod30DaysPeriodicAccrualProductWithAdvancedPaymentAllocationAndInterestRecalculation(
+                    (double) 80.0, rescheduleStrategyMethod);
+
+            CallFailedRuntimeException callFailedRuntimeException = Assertions.assertThrows(CallFailedRuntimeException.class,
+                    () -> loanProductHelper.createLoanProduct(loanProduct));
+
+            Assertions.assertTrue(callFailedRuntimeException.getMessage().contains("is not supported for Progressive loan schedule type"));
+        });
+    }
+
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:5140]: Would you mind to add a test case to cover whether the loan is cumulative but we are trying to set "Adjust last, unpaid period"?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: Done! updated

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionAccrualActivityPostingTest.java (hunk 5 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanTransactionAccrualActivityPostingTest.java]: Done! removed

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java (hunk 6 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:5111]: Done! TC extended

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java (hunk 7 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:5140]: Done! TC added

**Review by alberto-art3ch — COMMENTED** [review]
Changes applied, Thanks!

**Review by adamsaghy — APPROVED** [review]
LGTM
